Zend/zend_object_handlers.c

```diff
@@ -830,17 +868,31 @@ ZEND_API void zend_std_write_property(zval *object, zval *member, zval *value, v
 			}
 		}
 		if (EXPECTED(IS_VALID_PROPERTY_OFFSET(property_offset))) {
-			ZVAL_COPY_VALUE(OBJ_PROP(zobj, property_offset), value);
+			zend_property_info *prop_info;
+			zval tmp;
+
+			if (UNEXPECTED(prop_info = zend_object_fetch_property_type_info(Z_OBJCE_P(object), name, cache_slot))) {
+				ZVAL_COPY_VALUE(&tmp, value);
+				if (UNEXPECTED(!zend_verify_property_type(prop_info, &tmp, ZEND_CALL_USES_STRICT_TYPES(EG(current_execute_data))))) {
+					zend_verify_property_type_error(prop_info, value);
+					zval_ptr_dtor(value);
+					goto exit;
+				}
+				value = &tmp;
+			}
+			variable_ptr = OBJ_PROP(zobj, property_offset);
+			ZVAL_COPY_VALUE(variable_ptr, value);
```

Comment:
I'm not sure if we care, but something like this is going to leak:

```php
class A {
    public string $prop;
}

class B {
    public function __toString() {
        global $a;
        $a->prop = "dont ";
        $a->prop .= "leak ";
        $a->prop .= "me!";
        return "test";
    }
}

$a = new A;
$a->prop = new B;
var_dump($a);
```

Reply:
Quite weird, but should be fine to just simply add a zval_ptr_dtor() in if (prop_info) branch on success. Easy fix, so I care :-P